src/tiled/command.cpp

```diff
@@ -34,56 +34,58 @@ QString Command::finalWorkingDirectory() const
 {
     QString finalWorkingDirectory = workingDirectory;
 
-    // Perform variable replacement
-    if (Document *document = DocumentManager::instance()->currentDocument()) {
-        const QString fileName = document->fileName();
-
-        QFileInfo fileInfo(fileName);
-        QString mapPath = fileInfo.absolutePath();
-        finalWorkingDirectory.replace(
-            QLatin1String("%mappath"),
-            QString(QLatin1String("%1")).arg(mapPath));
-        // QProcess::setWorkingDirectory doesn't work with quoted paths
-
-        // TODO: %executablepath
-    }
-
-    return finalWorkingDirectory;
+    return replaceVariables(finalWorkingDirectory, true);
 }
 
 QString Command::finalCommand() const
 {
     QString finalCommand = QString(QLatin1String("%1 %2")).arg(executable).arg(arguments);
 
+    return replaceVariables(finalCommand);
+}
+
+QString Command::replaceVariables(const QString &string, bool flag) const
+{
+    QString finalString = string;
+
     // Perform variable replacement
     if (Document *document = DocumentManager::instance()->currentDocument()) {
         const QString fileName = document->fileName();
 
-        finalCommand.replace(QLatin1String("%mapfile"),
+        finalString.replace(QLatin1String("%mapfile"),
                              QString(QLatin1String("\"%1\"")).arg(fileName));
 
         QFileInfo fileInfo(fileName);
         QString mapPath = fileInfo.absolutePath();
-        finalCommand.replace(
-            QLatin1String("%mappath"),
-            QString(QLatin1String("\"%1\"")).arg(mapPath));
+
+        if (flag) {
+            finalString.replace(
+                QLatin1String("%mappath"),
+                QString(QLatin1String("%1")).arg(mapPath));
+        } else {
+            finalString.replace(
+                QLatin1String("%mappath"),
+                QString(QLatin1String("\"%1\"")).arg(mapPath));
+        }
```

Comment:
Yes, while merging into two functions I had the same doubt. What you suggest might work, I'm not sure though.